rcl/src/rcl/publisher.c

```diff
@@ -349,6 +349,29 @@ rcl_publisher_get_subscription_count(
   return RCL_RET_OK;
 }
 
+rmw_ret_t
+rcl_publisher_get_actual_qos(
+  const rcl_publisher_t * publisher,
+  rmw_qos_profile_t * qos)
+{
+  if (!rcl_publisher_is_valid(publisher)) {
```

Comment:
I'm not sure whether or not you need to do this, but you may need to use `rcl_publisher_is_valid_except_context` here instead, depending on how/where we use this function in rclcpp. Basically right now, this function will fail after `rclcpp::shutdown()`/`rcl_shutdown()` is called (or if SIGINT is received).

Reply:
Maybe the right thing to do is to leave it as-is and explicitly handle/suppress this issue, e.g.:

https://github.com/ros2/rclcpp/blob/b352d450319e84dd28aed78b800ab40938cec5c2/rclcpp/include/rclcpp/publisher.hpp#L350-L353

Since this is just an accessor (should be possible to calculate this during init) then it is maybe ok to access it during post-shutdown?

Reply:
The `get_options`, should (in my opinion) return the original options used. I could be convinced otherwise, but for now, that's what I think.

So in that light I'd say we still need `get_actual_qos`.

Why can't you just calculate the actual qos and store it separately (redundantly) during init?

Handling the error in rclcpp is possible, but only if it allows us to fail gracefully (not to be confused with silently). For instance, if we need to get the actual qos in order to decide the behavior of the publish function, you have to ask yourself what we'll do it it is unavailable, and if that mitigating action is ok.

Reply:
> Why can't you just calculate the actual qos and store it separately (redundantly) during init?

Sounds good to me, I will add that.